--- a/arch/arm/boot/dts/am33xx.dtsi	2015-01-19 10:30:29.110732635 -0600
+++ b/arch/arm/boot/dts/am33xx.dtsi	2015-01-19 10:33:25.002726093 -0600
@@ -638,8 +638,8 @@
 				#pwm-cells = <3>;
 				reg = <0x48300100 0x80>;
 				ti,hwmods = "ecap0";
-			clocks = <&l4ls_gclk>;
-			clock-names = "fck";
+				clocks = <&l4ls_gclk>;
+				clock-names = "fck";
 				status = "disabled";
 			};

@@ -648,8 +648,17 @@
 				#pwm-cells = <3>;
 				reg = <0x48300200 0x80>;
 				ti,hwmods = "ehrpwm0";
-			clocks = <&l4ls_gclk>;
-			clock-names = "fck";
+				clocks = <&l4ls_gclk>;
+				clock-names = "fck";
+				status = "disabled";
+			};
+
+			eqep0: eqep@0x48300180 {
+				compatible = "ti,am33xx-eqep";
+				reg = <0x48300180 0x80>;
+				interrupt-parent = <&intc>;
+				interrupts = <79>;
+				ti,hwmods = "eqep0";
 				status = "disabled";
 			};
 		};
@@ -672,8 +681,8 @@
 				#pwm-cells = <3>;
 				reg = <0x48302100 0x80>;
 				ti,hwmods = "ecap1";
-			clocks = <&l4ls_gclk>;
-			clock-names = "fck";
+				clocks = <&l4ls_gclk>;
+				clock-names = "fck";
 				status = "disabled";
 			};

@@ -682,8 +691,17 @@
 				#pwm-cells = <3>;
 				reg = <0x48302200 0x80>;
 				ti,hwmods = "ehrpwm1";
-			clocks = <&l4ls_gclk>;
-			clock-names = "fck";
+				clocks = <&l4ls_gclk>;
+				clock-names = "fck";
+				status = "disabled";
+			};
+
+			eqep1: eqep@0x48302180 {
+				compatible = "ti,am33xx-eqep";
+				reg = <0x48302180 0x80>;
+				interrupt-parent = <&intc>;
+				interrupts = <88>;
+				ti,hwmods = "eqep1";
 				status = "disabled";
 			};
 		};
@@ -706,8 +724,8 @@
 				#pwm-cells = <3>;
 				reg = <0x48304100 0x80>;
 				ti,hwmods = "ecap2";
-			clocks = <&l4ls_gclk>;
-			clock-names = "fck";
+				clocks = <&l4ls_gclk>;
+				clock-names = "fck";
 				status = "disabled";
 			};

@@ -716,8 +734,17 @@
 				#pwm-cells = <3>;
 				reg = <0x48304200 0x80>;
 				ti,hwmods = "ehrpwm2";
-			clocks = <&l4ls_gclk>;
-			clock-names = "fck";
+				clocks = <&l4ls_gclk>;
+				clock-names = "fck";
+				status = "disabled";
+			};
+
+			eqep2: eqep@0x48304180 {
+				compatible = "ti,am33xx-eqep";
+				reg = <0x48304180 0x80>;
+				interrupt-parent = <&intc>;
+				interrupts = <89>;
+				ti,hwmods = "eqep2";
 				status = "disabled";
 			};
 		};
